.docs/architecture/adr-001-general-parser-architecture.md

```diff
@@ -0,0 +1,234 @@
+# ADR-001: General parser architecture
+
+
+## Changelog
+
+- March 11th, 2024: Init draft;
+- March 13th, 2024: First draft;    
+
+## Status
+
+PROPOSED
+
+## Abstract
+
+This ADR outlines the general parser architecture designed for Juno, addressing the need for enhanced interfaces to support various chains.
+
+## Context
+
+Juno currently lacks the necessary architecture documents and interfaces to effectively support multiple chains. Additionally, it faces challenges in handling messages and transactions, especially when dealing with different blockchain protocols and versions. There is a pressing need to establish a robust and adaptable architecture that can accommodate the diverse requirements of various chains while maintaining efficiency and reliability. 
+
+## Decision
+
+To address these challenges, the decision is made to redesign Juno's parser architecture. The architecture will prioritize modularity and extensibility, allowing seamless integration with different blockchain networks and protocols. Key components such as enqueuers, workers, and modules will be defined with clear responsibilities and interfaces to facilitate easy extension and customization. Moreover, the architecture will incorporate mechanisms to handle transactions, messages, and other blockchain operations efficiently.
+ 
+The provided image offer a comprehensive overview of the Juno parser, detailing its core components and functionalities.
+
+![Architecture](../.img/adr-001-general-architecture.png)
+
+Let's break down each element briefly by following topics.
+
+### Context
+
+The Context interface acts as a central hub shared among all elements within the parser, including enqueuers, workers, and other components.
+
+```go
+type Context interface {
+    // Context represents the manager for the concurrent jobs
+    context.Context
+
+    // Database returns the database used to operate blocks by workers
+    Database() Database
```

Comment:
Is there a reason not to use a singleton for the Database client?

Reply:
Yeah, we should use singleton on Context, Node and Database, inherited the previous design. 

In golang, Singleton is in the implementation level, interface just a way to abstract the methods.
https://thedevelopercafe.com/articles/singleton-in-golang-839d8610958b

Reply:
Oh understood 👍.

Why do we need the `Context` interface though? Can't we just use a function to get the database client like in the article you linked?

Reply:
`Context` is mainly a concept to manage concurrency program, and it is shared to every goroutine tasks. Then we can manage all the workers in a easy way. [Here](https://reintech.io/blog/introduction-to-go-context-package) is the brief reference.

In addition, putting Database, Node and etc. it follows the original concept from current Juno that including the shared singleton objects among workers.

Do you think we use `Start(context.Context, Database, Node)` would be better?

Reply:
I don't think `Start(context.Context, Database, Node)` is any better 😄.

I suggested retrieving the database client by the Singleton getter method like `GetCache` in the article you linked. I mean from within the function body.

But I think your approach is the recommended one using Golang. I am more used to doing things like that https://medium.com/learnfazz/flawed-singleton-pattern-for-mocking-go-e90794891faf.

Again I think you have the right approach 👍.